### ISSUE TITLE
retry db reads and idempotent writes

### ISSUE THREAD

**Issue by gurayAlsac** [google/exposure-notifications-server#652]
### TL;DR
We should retry db operations that are safe to retry (like config reads, or idempotent writes).

### Design

**Proposal**
There are cases of transient db connection errors that could be retried. E.g., if we fail to read AppConfig to see if an exposure upload was from an authorized app, we could drop keys on the floor:
https://github.com/google/exposure-notifications-server/blob/76f3f89c7415db09ef0aa302305fb9279f221bfb/internal/publish/publish.go#L120

It looks like we have some retry logic in the db connections on startup and could maybe make use of that in a steady state:
https://github.com/google/exposure-notifications-server/blob/1f63d1942c08a03bb48a5b668a0ccbf4dd6c39b0/internal/database/database_util.go#L116

**Alternatives considered**
We could rely on clients to retry, but there's no harm in being flexible in both layers and handling transient errors in both client and the server.

**Comment by sethvargo** [comment]
I'm definitely down to retry - we have a library that makes it easy. I'm interested in learning the situations where failures happen - do you have server logs? The most likely candidate is during a scale down event, in which case a server-side retry wouldn't help.

**Comment by mikehelmick** [comment]
Thank I saw the logs - this was a connection reset. I think retry would have helped.

Cloud run won't scale down during live requests

**Comment by sethvargo** [comment]
Per https://github.com/jackc/pgx/issues/74#issuecomment-139832623, it looks like we just need to wrap reads in serializable transactions.

/assign
